File: backend/app/api/routers/relationships.py
import uuid
from typing import Annotated, Literal

from fastapi import APIRouter, Query, Response, status

from app.api.dependencies import RelationshipContextDependency
from app.api.schemas.relationships import (
    ConfirmCandidateRequest,
    DetectionResponse,
    ManualRelationshipRequest,
    PolymorphicMappingRequest,
    PolymorphicRelationshipRequest,
    PolymorphicRelationshipResponse,
    RelationshipGraphResponse,
    RelationshipListResponse,
    RelationshipUpdateRequest,
    UnifiedRelationshipResponse,
)
from app.application.relationships import (
    ConfirmCandidateService,
    CreateManualRelationshipService,
    DeleteRelationshipService,
    DetectRelationshipCandidatesService,
    ListRelationshipsService,
    PolymorphicRelationshipService,
    RejectCandidateService,
    RelationshipGraphService,
    SetRelationshipEnabledService,
    UpdateRelationshipService,
)
from app.domain.connections.errors import PublicError

router = APIRouter(
    prefix="/connections/{connection_id}/relationships",
    tags=["relationships"],
)


@router.get("", response_model=RelationshipListResponse)
async def list_relationships(
    connection_id: uuid.UUID,
    context: RelationshipContextDependency,
    relationship_type: Literal["physical", "inferred", "manual", "polymorphic"] | None = None,
    status_filter: Annotated[str | None, Query(alias="status")] = None,
) -> RelationshipListResponse:
    return await ListRelationshipsService(context).execute(
        connection_id,
        relationship_type=relationship_type,
        status=status_filter,
    )


@router.get("/graph", response_model=RelationshipGraphResponse)
async def relationship_graph(
    connection_id: uuid.UUID, context: RelationshipContextDependency
) -> RelationshipGraphResponse:
    return await RelationshipGraphService(context).execute(connection_id)


@router.get("/candidates", response_model=RelationshipListResponse)
async def list_candidates(
    connection_id: uuid.UUID, context: RelationshipContextDependency
) -> RelationshipListResponse:
    return await ListRelationshipsService(context).execute(connection_id, status="suggested")


@router.post("/detect", response_model=DetectionResponse)
async def detect_candidates(
    connection_id: uuid.UUID, context: RelationshipContextDependency
) -> DetectionResponse:
    return await DetectRelationshipCandidatesService(context).execute(connection_id)


@router.post(
    "/candidates/{candidate_id}/confirm",
    response_model=UnifiedRelationshipResponse,
)
async def confirm_candidate(
    connection_id: uuid.UUID,
    candidate_id: uuid.UUID,
    request: ConfirmCandidateRequest,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await ConfirmCandidateService(context).execute(connection_id, candidate_id, request)


@router.post(
    "/candidates/{candidate_id}/reject",
    response_model=UnifiedRelationshipResponse,
)
async def reject_candidate(
    connection_id: uuid.UUID,
    candidate_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await RejectCandidateService(context).execute(connection_id, candidate_id)


@router.post(
    "/manual",
    response_model=UnifiedRelationshipResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_manual_relationship(
    connection_id: uuid.UUID,
    request: ManualRelationshipRequest,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await CreateManualRelationshipService(context).execute(connection_id, request)


@router.post(
    "/polymorphic",
    response_model=PolymorphicRelationshipResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_polymorphic_relationship(
    connection_id: uuid.UUID,
    request: PolymorphicRelationshipRequest,
    context: RelationshipContextDependency,
) -> PolymorphicRelationshipResponse:
    return await PolymorphicRelationshipService(context).create(connection_id, request)


@router.get(
    "/polymorphic/{relationship_id}",
    response_model=PolymorphicRelationshipResponse,
)
async def get_polymorphic_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> PolymorphicRelationshipResponse:
    return await PolymorphicRelationshipService(context).get(connection_id, relationship_id)


@router.patch(
    "/polymorphic/{relationship_id}",
    response_model=PolymorphicRelationshipResponse,
)
async def update_polymorphic_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    request: RelationshipUpdateRequest,
    context: RelationshipContextDependency,
) -> PolymorphicRelationshipResponse:
    relation = await context.catalog.polymorphic(connection_id, relationship_id)
    if relation is None:
        raise PublicError("RELATIONSHIP_NOT_FOUND", "La relación no existe.", 404)
    for name, value in request.model_dump(exclude_unset=True).items():
        if name in {"display_name", "description"}:
            setattr(relation, name, value)
    await context.session.commit()
    return await PolymorphicRelationshipService(context).get(connection_id, relationship_id)


@router.delete("/polymorphic/{relationship_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_polymorphic_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> Response:
    await PolymorphicRelationshipService(context).delete(connection_id, relationship_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/polymorphic/{relationship_id}/mappings",
    response_model=PolymorphicRelationshipResponse,
)
async def add_polymorphic_mapping(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    request: PolymorphicMappingRequest,
    context: RelationshipContextDependency,
) -> PolymorphicRelationshipResponse:
    return await PolymorphicRelationshipService(context).add_mapping(
        connection_id, relationship_id, request
    )


@router.patch(
    "/polymorphic/{relationship_id}/mappings/{mapping_id}",
    response_model=PolymorphicRelationshipResponse,
)
async def update_polymorphic_mapping(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    mapping_id: uuid.UUID,
    request: PolymorphicMappingRequest,
    context: RelationshipContextDependency,
) -> PolymorphicRelationshipResponse:
    return await PolymorphicRelationshipService(context).update_mapping(
        connection_id, relationship_id, mapping_id, request
    )


@router.delete(
    "/polymorphic/{relationship_id}/mappings/{mapping_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_polymorphic_mapping(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    mapping_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> Response:
    await PolymorphicRelationshipService(context).delete_mapping(
        connection_id, relationship_id, mapping_id
    )
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/polymorphic/{relationship_id}/discover-values")
async def discover_polymorphic_values(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> None:
    del connection_id, relationship_id
    if not context.settings.ENABLE_POLYMORPHIC_VALUE_DISCOVERY:
        raise PublicError(
            "POLYMORPHIC_DISCOVERY_DISABLED",
            "El descubrimiento de valores polimórficos está deshabilitado.",
            403,
        )
    raise PublicError(
        "POLYMORPHIC_DISCOVERY_DISABLED",
        "El descubrimiento remoto no está habilitado en esta fase.",
        501,
    )


@router.get("/{relationship_id}", response_model=UnifiedRelationshipResponse)
async def get_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    response = await ListRelationshipsService(context).execute(connection_id)
    for item in response.items:
        if item.id == relationship_id:
            return item
    raise PublicError("RELATIONSHIP_NOT_FOUND", "La relación no existe.", 404)


@router.patch("/{relationship_id}", response_model=UnifiedRelationshipResponse)
async def update_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    request: RelationshipUpdateRequest,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await UpdateRelationshipService(context).execute(connection_id, relationship_id, request)


@router.delete("/{relationship_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> Response:
    await DeleteRelationshipService(context).execute(connection_id, relationship_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{relationship_id}/disable", response_model=UnifiedRelationshipResponse)
async def disable_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await SetRelationshipEnabledService(context).execute(
        connection_id, relationship_id, False
    )


@router.post("/{relationship_id}/enable", response_model=UnifiedRelationshipResponse)
async def enable_relationship(
    connection_id: uuid.UUID,
    relationship_id: uuid.UUID,
    context: RelationshipContextDependency,
) -> UnifiedRelationshipResponse:
    return await SetRelationshipEnabledService(context).execute(
        connection_id, relationship_id, True
    )

File: frontend/src/services/relationships.ts
import type {
  RelationshipGraph,
  RelationshipList,
  SemanticEntity,
  SemanticField,
  UnifiedRelationship,
} from "../features/relationships/types";
import { apiRequest } from "./shared";

const base = (connectionId: string) =>
  `/connections/${encodeURIComponent(connectionId)}/relationships`;
const semanticBase = (connectionId: string) =>
  `/connections/${encodeURIComponent(connectionId)}/semantic`;

export const listRelationships = (connectionId: string) =>
  apiRequest<RelationshipList>(base(connectionId));
export const getRelationshipGraph = (connectionId: string) =>
  apiRequest<RelationshipGraph>(`${base(connectionId)}/graph`);
export const listRelationshipCandidates = (connectionId: string) =>
  apiRequest<RelationshipList>(`${base(connectionId)}/candidates`);
export const detectRelationshipCandidates = (connectionId: string) =>
  apiRequest<{ detected: number; created: number; preserved_rejections: number }>(
    `${base(connectionId)}/detect`,
    { method: "POST" },
  );
export const confirmRelationshipCandidate = (
  connectionId: string,
  candidateId: string,
  payload: Record<string, unknown> = {},
) =>
  apiRequest<UnifiedRelationship>(
    `${base(connectionId)}/candidates/${encodeURIComponent(candidateId)}/confirm`,
    { method: "POST", body: JSON.stringify(payload) },
  );
export const rejectRelationshipCandidate = (
  connectionId: string,
  candidateId: string,
) =>
  apiRequest<UnifiedRelationship>(
    `${base(connectionId)}/candidates/${encodeURIComponent(candidateId)}/reject`,
    { method: "POST" },
  );
export const createManualRelationship = (
  connectionId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<UnifiedRelationship>(`${base(connectionId)}/manual`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
export const updateRelationship = (
  connectionId: string,
  relationshipId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<UnifiedRelationship>(
    `${base(connectionId)}/${encodeURIComponent(relationshipId)}`,
    { method: "PATCH", body: JSON.stringify(payload) },
  );
export const deleteRelationship = (connectionId: string, relationshipId: string) =>
  apiRequest<unknown>(`${base(connectionId)}/${encodeURIComponent(relationshipId)}`, {
    method: "DELETE",
  });
export const enableRelationship = (connectionId: string, relationshipId: string) =>
  apiRequest<UnifiedRelationship>(
    `${base(connectionId)}/${encodeURIComponent(relationshipId)}/enable`,
    { method: "POST" },
  );
export const disableRelationship = (connectionId: string, relationshipId: string) =>
  apiRequest<UnifiedRelationship>(
    `${base(connectionId)}/${encodeURIComponent(relationshipId)}/disable`,
    { method: "POST" },
  );
export const createPolymorphicRelationship = (
  connectionId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<Record<string, unknown>>(`${base(connectionId)}/polymorphic`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
export const getPolymorphicRelationship = (connectionId: string, relationshipId: string) =>
  apiRequest<{
    id: string; source_entity_id: string; source_entity: string; type_field_id: string;
    type_field: string; id_field_id: string; id_field: string; name: string;
    display_name: string; status: string; is_enabled: boolean; invalid_reason: string | null;
    mappings: Array<{ id: string; type_value: string; target_entity_id: string; target_entity: string; target_field_id: string; target_field: string; display_name: string; is_enabled: boolean }>;
  }>(`${base(connectionId)}/polymorphic/${encodeURIComponent(relationshipId)}`);
export const addPolymorphicMapping = (
  connectionId: string,
  relationshipId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<Record<string, unknown>>(
    `${base(connectionId)}/polymorphic/${encodeURIComponent(relationshipId)}/mappings`,
    { method: "POST", body: JSON.stringify(payload) },
  );
export const updatePolymorphicMapping = (
  connectionId: string,
  relationshipId: string,
  mappingId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<Record<string, unknown>>(
    `${base(connectionId)}/polymorphic/${encodeURIComponent(relationshipId)}/mappings/${encodeURIComponent(mappingId)}`,
    { method: "PATCH", body: JSON.stringify(payload) },
  );
export const deletePolymorphicMapping = (
  connectionId: string,
  relationshipId: string,
  mappingId: string,
) =>
  apiRequest<unknown>(
    `${base(connectionId)}/polymorphic/${encodeURIComponent(relationshipId)}/mappings/${encodeURIComponent(mappingId)}`,
    { method: "DELETE" },
  );
export const listSemanticEntities = (connectionId: string) =>
  apiRequest<{ items: SemanticEntity[]; total: number }>(
    `${semanticBase(connectionId)}/entities`,
  );
export const getSemanticEntity = (connectionId: string, entityId: string) =>
  apiRequest<SemanticEntity>(
    `${semanticBase(connectionId)}/entities/${encodeURIComponent(entityId)}`,
  );
export const updateSemanticEntity = (
  connectionId: string,
  entityId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<SemanticEntity>(
    `${semanticBase(connectionId)}/entities/${encodeURIComponent(entityId)}`,
    { method: "PATCH", body: JSON.stringify(payload) },
  );
export const updateSemanticField = (
  connectionId: string,
  fieldId: string,
  payload: Record<string, unknown>,
) =>
  apiRequest<SemanticField>(
    `${semanticBase(connectionId)}/fields/${encodeURIComponent(fieldId)}`,
    { method: "PATCH", body: JSON.stringify(payload) },
  );
